src/artifact/decl.rs

```diff
@@ -29,15 +28,145 @@ impl ImportKind {
     }
 }
 
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+/// Linker binding scope of a definition
+pub enum Scope {
+    /// Available to all components
+    Global,
+    /// Available only inside the defining component
+    Local,
+    /// Available to all modules, but only selected if a Global
+    /// definition is not found
+    Weak,
+}
+
+macro_rules! scope_methods {
+    () => {
+    /// Set scope to global
+    pub fn global(self) -> Self {
+        self.with_scope(Scope::Global)
+    }
+    /// Set scope to local
+    pub fn local(self) -> Self {
+        self.with_scope(Scope::Local)
+    }
+    /// Set scope to weak
+    pub fn weak(self) -> Self {
+        self.with_scope(Scope::Weak)
+    }
+    /// Builder for scope
+    pub fn with_scope(mut self, scope: Scope) -> Self {
+        self.scope = scope;
+        self
+    }
+    /// Gst scope
+    pub fn get_scope(&self) -> Scope {
+        self.scope
+    }
+    /// Set scope
+    pub fn set_scope(&mut self, scope: Scope) {
+        self.scope = scope;
+    }
+    /// Check if scope is `Scope::Global`
+    pub fn is_global(&self) -> bool {
+        self.scope == Scope::Global
+    }
+}}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+/// Linker visibility of a definition
+pub enum Visibility {
+    /// Visibility determined by the symbol's `Scope`.
+    Default,
+    /// Visible in other components, but cannot be preempted. References to the symbol must be
+    /// resolved to this definition in that component, even if another definition would interpose
+    /// by the default rules.
+    Protected,
+    /// Not visible to other components, plus the constraints provided by `Protected`.
+    Hidden,
+}
+
+macro_rules! visibility_methods {
+    () => {
+    /// Set visibility to default
+    pub fn default_visibility(self) -> Self {
+        self.with_visibility(Visibility::Default)
+    }
+    /// Set visibility to protected
+    pub fn protected(self) -> Self {
+        self.with_visibility(Visibility::Protected)
+    }
+    /// Set visibility to hidden
+    pub fn hidden(self) -> Self {
+        self.with_visibility(Visibility::Hidden)
+    }
+    /// Builder for visibility
+    pub fn with_visibility(mut self, visibility: Visibility) -> Self {
+        self.visibility =visibility;
+        self
+    }
+    /// Get visibility
+    pub fn get_visibility(&self) -> Visibility {
+        self.visibility
+    }
+    /// Set visibility
+    pub fn set_visibility(&mut self, visibility: Visibility) {
+        self.visibility = visibility;
+    }
+}}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+/// Type of data declared
+pub enum DataType {
+    /// Ordinary raw bytes
+    Bytes,
+    /// 0-terminated ascii string
```

Comment:
Technically 0-terminated byte sequences could be things other than ASCII. I suggest just saying "C-style" here.